File: Src/IronJS/Legacy.fs
namespace IronJS.Legacy

// This file contains support classes needed in IronJS 
// that are not available on Mono and .NET version that 
// are earlier then 4.0

open System.Collections
open System.Collections.Generic

#if LEGACY_HASHSET
type HashSet<'a when 'a : equality>() =
  let o = obj()
  let storage = new Dictionary<'a, obj>();

  member x.Add(item) = (x :> ICollection<'a>).Add(item)
  member x.Clear() = (x :> ICollection<'a>).Clear()
  member x.Contains(item) = (x :> ICollection<'a>).Contains(item)
  member x.CopyTo(a, b) = (x :> ICollection<'a>).CopyTo(a, b)
  member x.Count = (x :> ICollection<'a>).Count
  member x.IsReadOnly = (x :> ICollection<'a>).IsReadOnly
  member x.GetEnumerator() : IEnumerator<'a> = (x :> ICollection<'a>).GetEnumerator()

  interface ICollection<'a> with
    member x.Add(item) = storage.[item] <- o;
    member x.Clear() = storage.Clear();
    member x.Contains(item) = storage.ContainsKey(item);
    member x.CopyTo(_:'a array,_:int) : unit = failwith "Not implemented"
    member x.Count = storage.Count
    member x.IsReadOnly = false
    member x.Remove(item) = storage.Remove(item)

    member x.GetEnumerator() : IEnumerator<'a> = 
     (seq { for x in storage do yield x.Key }).GetEnumerator()

    member x.GetEnumerator() : IEnumerator = 
     (seq { for x in storage do yield x.Key }).GetEnumerator() :> IEnumerator
#endif

#if LEGACY_DELEGATES
type Action = delegate of unit -> unit
type Action<'a, 'b> = delegate of 'a * 'b -> unit
type Action<'a, 'b, 'c> = delegate of 'a * 'b * 'c -> unit
type Action<'a, 'b, 'c, 'd> = delegate of 'a * 'b * 'c * 'd -> unit

type Func<'r> = delegate of unit -> 'r
type Func<'a, 'r> = delegate of 'a -> 'r
type Func<'a, 'b, 'r> = delegate of 'a * 'b -> 'r
type Func<'a, 'b, 'c,  'r> = delegate of 'a * 'b * 'c -> 'r
type Func<'a, 'b, 'c, 'd, 'r> = delegate of 'a * 'b * 'c * 'd -> 'r
#endif

#if LEGACY_DELEGATES_HIGH_ARITY
type Func<'a, 'b, 'c, 'd, 'e, 'r> = delegate of 'a * 'b * 'c * 'd * 'e -> 'r
type Func<'a, 'b, 'c, 'd, 'e, 'f, 'r> = delegate of 'a * 'b * 'c * 'd * 'e * 'f -> 'r
#endif

#if LEGACY_SORTED_DICT
type SortedDictionary<'k, 'v>() =
  class
    
  end
#endif

#if LEGACY_BIGINT_TRYPARSE
type BigIntegerParser() =
  
  static member TryParse(s, f, i, bi:bigint byref) =
    try
      bi <- System.Numerics.BigInteger.Parse(s)
      true

    with
      | _ -> false
#endif

File: Src/IronJS.Runtime/Legacy.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace IronJS.Runtime
{
#if LEGACY_HASHSET
    public class HashSet<T> : ICollection<T>
    {
        object slug = new object();
        Dictionary<T, object> storage = new Dictionary<T, object>();

        public void Add(T item)
        {
            this.storage[item] = slug;
        }

        public void Clear()
        {
            this.storage.Clear();
        }

        public bool Contains(T item)
        {
            return this.storage.ContainsKey(item);
        }

        public void CopyTo(T[] array, int arrayIndex)
        {
            this.storage.Keys.CopyTo(array, arrayIndex);
        }

        public int Count
        {
            get { return this.storage.Count; }
        }

        public bool IsReadOnly
        {
            get { return false; }
        }

        public bool Remove(T item)
        {
            return this.storage.Remove(item);
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this.storage.Keys.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
#endif

#if LEGACY_DELEGATES
    public delegate void Action();
    public delegate void Action<in T>(T obj);
    public delegate void Action<in T1, in T2>(T1 arg1, T2 arg2);
    public delegate void Action<in T1, in T2, in T3>(T1 arg1, T2 arg2, T3 arg3);

    public delegate TResult Func<out TResult>();
    public delegate TResult Func<in T, out TResult>(T arg);
    public delegate TResult Func<in T1, in T2, out TResult>(T1 arg1, T2 arg2);
    public delegate TResult Func<in T1, in T2, in T3, out TResult>(T1 arg1, T2 arg2, T3 arg3);
    public delegate TResult Func<in T1, in T2, in T3, in T4, out TResult>(T1 arg1, T2 arg2, T3 arg3, T4 arg4);
#endif

#if LEGACY_DELEGATES_HIGH_ARITY
    public delegate TResult Func<in T1, in T2, in T3, in T4, in T5, out TResult>(T1 arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5);
    public delegate TResult Func<in T1, in T2, in T3, in T4, in T5, in T6, out TResult>(T1 arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5, T6 arg6);
#endif

#if LEGACY_SORTED_DICT
    public class SortedDictionary<TKey, TValue>
    {
    }
#endif

#if LEGACY_BIGINT
    public static class BigIntegerParser
    {
        public static bool TryParse(string value, out System.Numerics.BigInteger result)
        {
            return TryParse(value, System.Globalization.NumberStyles.Any, System.Globalization.CultureInfo.InvariantCulture, out result);
        }

        public static bool TryParse(string value, System.Globalization.NumberStyles style, System.Globalization.CultureInfo culture, out System.Numerics.BigInteger result)
        {
            try
            {
                result = System.Numerics.BigInteger.Parse(value);
                return true;
            }
            catch (FormatException)
            {
                result = System.Numerics.BigInteger.Zero;
                return false;
            }
        }
    }
#endif
}
